services/cache2/provider.go

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package cache2
+
+import "time"
+
+// Provider is a provider for Cache
+type Provider interface {
+	// NewCache creates a new cache with given size.
+	NewCache(size int) Cache
+	// NewCacheWithParams creates a new cache with the given parameters.
+	NewCacheWithParams(size int, name string, defaultExpiryInSecs int, invalidateClusterEvent string) Cache
```

Comment:
on second thought, i am thinking to change this API a little bit, with a slight drawback of a dozens of more line updates. Please let me know what you think before I proceed:

```
// type ProviderOptions {
//   size int
//   defaultExpiry time.Duration
//   name string
//   invalidateClusterEvent string
// }

NewCacheWithParams(opt *ProviderOptions) Cache
```

the rationale is as follow:

1. Having a struct `ProviderOptions` so that we can keep the same interface in case a new field is required in the future. e.g. the `name` field is needed for a redis cache but not for a LRU cache. What if some other fields are needed for a new type of cache we support in future?

2. change `defaultExpiry` to a `time.Duration`. This one is pretty obvious.

Reply:
Thanks. This is on my list. I will take a look whenever I get some time.

Reply:
Ok, your changes to ProviderOptions look reasonable to me. Since we are changing the interface, let's change it to be better. I am okay with changing more code if that gives us more benefit in the long term. Let's also change the expiry to time.Duration.

Did you want to update the references from using the LRU implementation directly to going via the provider in another PR or this PR itself ? I am okay with both options.

Reply:
Yes, I agree with this too. Is better to change it now, and with 4 parameters that are all options, I think is less error prone to have a data structure for that.

Reply:
Thanks. I will update this CR.

> Did you want to update the references from using the LRU implementation directly to going via the provider in another PR or this PR itself ? I am okay with both options.

My goal was to migrate usage as it is with minimum logical changes possible. I don't plan to migrate direct LRU references to cacheProvider. This PR contains all direct references to LRU and the rest of the code base all uses cacheProvider. 

I think we can do migration of direct LRU usage as a separate topic, or as a followup after the we are fully migrated to cache2 interface.

Reply:
Sorry could you explain your comment a bit ? I am not clear what you meant by these lines

> I don't plan to migrate direct LRU references to cacheProvider.

> I think we can do migration of direct LRU usage as a separate topic, 

These two seem to be conflicting to me.

Reply:
Sorry for being confusing. I meant to say that the first goal in my mind is to complete `cache2` package migration. Other improvements such as removing direct LRU references can be done at later stages, or do you think it is better to do it now?

Reply:
Got it. Yes, I agree with that.